File: src/dere_daemon/routers/notifications.py
"""Notification system endpoints."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from fastapi import APIRouter, Depends
from loguru import logger
from pydantic import BaseModel
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from dere_daemon.dependencies import get_db
from dere_shared.models import Notification, NotificationContext

router = APIRouter(prefix="/notifications", tags=["notifications"])


# Request/Response models
class NotificationCreateRequest(BaseModel):
    user_id: str
    target_medium: str
    target_location: str
    message: str
    priority: str
    routing_reasoning: str
    parent_notification_id: int | None = None
    context_snapshot: dict[str, Any] | None = None
    trigger_type: str | None = None
    trigger_id: str | None = None
    trigger_data: dict[str, Any] | None = None


class NotificationDeliveredRequest(BaseModel):
    notification_id: int


class NotificationFailedRequest(BaseModel):
    notification_id: int
    error_message: str


class NotificationQueryRequest(BaseModel):
    user_id: str
    since: str


@router.post("/create")
async def notifications_create(req: NotificationCreateRequest, db: AsyncSession = Depends(get_db)):
    """Create a notification in the queue for delivery.

    Called by ambient monitor when it decides to engage.
    """
    notification = Notification(
        user_id=req.user_id,
        target_medium=req.target_medium,
        target_location=req.target_location,
        message=req.message,
        priority=req.priority,
        routing_reasoning=req.routing_reasoning,
        status="pending",
        created_at=datetime.now(UTC),
        parent_notification_id=req.parent_notification_id,
    )
    db.add(notification)
    await db.flush()

    if req.context_snapshot or req.trigger_type:
        notif_context = NotificationContext(
            notification_id=notification.id,
            trigger_type=req.trigger_type,
            trigger_id=req.trigger_id,
            trigger_data=req.trigger_data,
            context_snapshot=req.context_snapshot,
        )
        db.add(notif_context)

    await db.commit()

    # Truncate message for logging
    message_preview = req.message[:100] + "..." if len(req.message) > 100 else req.message
    logger.info(
        "Notification {} created: {} -> {} ({}) | \"{}\"",
        notification.id,
        req.target_medium,
        req.target_location,
        req.priority,
        message_preview,
    )
    return {"notification_id": notification.id, "status": "queued"}


@router.get("/recent")
async def notifications_recent(
    user_id: str,
    limit: int = 5,
    db: AsyncSession = Depends(get_db),
):
    """Get recent notifications for a user (for ambient monitor context)."""
    stmt = (
        select(Notification)
        .where(Notification.user_id == user_id)
        .order_by(Notification.created_at.desc())
        .limit(limit)
    )
    result = await db.execute(stmt)
    notifications = result.scalars().all()

    return {
        "notifications": [
            {
                "id": n.id,
                "message": n.message,
                "priority": n.priority,
                "status": n.status,
                "created_at": n.created_at.isoformat() if n.created_at else None,
                "delivered_at": n.delivered_at.isoformat() if n.delivered_at else None,
            }
            for n in notifications
        ]
    }


@router.get("/pending")
async def notifications_pending(medium: str, db: AsyncSession = Depends(get_db)):
    """Get pending notifications for a specific medium.

    Bots poll this endpoint to retrieve messages that need to be delivered.
    """
    stmt = (
        select(Notification)
        .where(Notification.target_medium == medium)
        .where(Notification.status == "pending")
        .order_by(Notification.priority.desc(), Notification.created_at.asc())
    )
    result = await db.execute(stmt)
    notifications = result.scalars().all()

    return {
        "notifications": [
            {
                "id": n.id,
                "user_id": n.user_id,
                "target_location": n.target_location,
                "message": n.message,
                "priority": n.priority,
                "routing_reasoning": n.routing_reasoning,
                "created_at": n.created_at,
            }
            for n in notifications
        ]
    }


@router.post("/recent_unacknowledged")
async def notifications_recent_unacknowledged(
    req: NotificationQueryRequest, db: AsyncSession = Depends(get_db)
):
    """Query recent unacknowledged notifications for escalation context.

    Returns notifications that were delivered but not acknowledged within the lookback period.
    """
    since_time = datetime.fromisoformat(req.since)

    stmt = (
        select(Notification)
        .where(
            Notification.user_id == req.user_id,
            Notification.created_at >= since_time,
            Notification.status == "delivered",
            ~Notification.acknowledged,
        )
        .order_by(Notification.created_at.desc())
        .limit(10)
    )

    result = await db.execute(stmt)
    notifications = result.scalars().all()

    return {
        "notifications": [
            {
                "id": n.id,
                "message": n.message,
                "priority": n.priority,
                "created_at": n.created_at.isoformat() if n.created_at else None,
                "delivered_at": n.delivered_at.isoformat() if n.delivered_at else None,
                "status": n.status,
                "acknowledged": n.acknowledged,
                "parent_notification_id": n.parent_notification_id,
            }
            for n in notifications
        ]
    }


@router.post("/{notification_id}/delivered")
async def notification_delivered(notification_id: int, db: AsyncSession = Depends(get_db)):
    """Mark a notification as successfully delivered.

    Called by bots after successfully sending a message.
    """
    stmt = (
        update(Notification)
        .where(Notification.id == notification_id)
        .values(
            status="delivered",
            delivered_at=datetime.now(UTC),
        )
    )
    await db.execute(stmt)
    await db.commit()

    logger.info("Notification {} marked as delivered", notification_id)
    return {"status": "delivered"}


@router.post("/{notification_id}/acknowledge")
async def notification_acknowledge(notification_id: int, db: AsyncSession = Depends(get_db)):
    """Mark a notification as acknowledged by the user.

    Called when user responds/interacts after receiving notification.
    This prevents escalation of the notification.
    """
    stmt = (
        update(Notification)
        .where(Notification.id == notification_id)
        .values(
            acknowledged=True,
            acknowledged_at=datetime.now(UTC),
        )
    )
    await db.execute(stmt)
    await db.commit()

    logger.info("Notification {} acknowledged by user", notification_id)
    return {"status": "acknowledged"}


@router.post("/{notification_id}/failed")
async def notification_failed(
    notification_id: int, req: NotificationFailedRequest, db: AsyncSession = Depends(get_db)
):
    """Mark a notification as failed with error message.

    Called by bots when message delivery fails.
    """
    stmt = (
        update(Notification)
        .where(Notification.id == notification_id)
        .values(
            status="failed",
            error_message=req.error_message,
            delivered_at=datetime.now(UTC),
        )
    )
    await db.execute(stmt)
    await db.commit()

    logger.warning("Notification {} failed: {}", notification_id, req.error_message)
    return {"status": "failed"}

File: packages/daemon/src/routes/notifications.ts
import type { Hono } from "hono";
import { getDb } from "../db.js";

function nowDate(): Date {
  return new Date();
}

async function parseJson<T>(req: Request): Promise<T | null> {
  try {
    return (await req.json()) as T;
  } catch {
    return null;
  }
}

function toJsonRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

export function registerNotificationRoutes(app: Hono): void {
  app.post("/notifications/create", async (c) => {
    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    if (!payload) {
      return c.json({ error: "Invalid JSON payload" }, 400);
    }

    const userId = payload.user_id;
    const targetMedium = payload.target_medium;
    const targetLocation = payload.target_location;
    const message = payload.message;
    const priority = payload.priority;
    const routingReasoning = payload.routing_reasoning;

    if (
      typeof userId !== "string" ||
      typeof targetMedium !== "string" ||
      typeof targetLocation !== "string" ||
      typeof message !== "string" ||
      typeof priority !== "string" ||
      typeof routingReasoning !== "string"
    ) {
      return c.json({ error: "Missing required fields" }, 400);
    }

    const db = await getDb();
    const now = nowDate();

    const notification = await db
      .insertInto("ambient_notifications")
      .values({
        user_id: userId,
        target_medium: targetMedium,
        target_location: targetLocation,
        message,
        priority,
        routing_reasoning: routingReasoning,
        status: "pending",
        created_at: now,
        delivered_at: null,
        parent_notification_id:
          typeof payload.parent_notification_id === "number"
            ? payload.parent_notification_id
            : null,
        acknowledged: false,
        acknowledged_at: null,
        response_time: null,
        error_message: null,
      })
      .returning(["id"])
      .executeTakeFirstOrThrow();

    if (payload.context_snapshot || payload.trigger_type) {
      await db
        .insertInto("notification_context")
        .values({
          notification_id: notification.id,
          trigger_type: typeof payload.trigger_type === "string" ? payload.trigger_type : null,
          trigger_id: typeof payload.trigger_id === "string" ? payload.trigger_id : null,
          trigger_data: toJsonRecord(payload.trigger_data),
          context_snapshot: toJsonRecord(payload.context_snapshot),
          created_at: now,
        })
        .execute();
    }

    return c.json({ notification_id: notification.id, status: "queued" });
  });

  app.get("/notifications/recent", async (c) => {
    const userId = c.req.query("user_id");
    const limit = Number(c.req.query("limit") ?? 5);
    if (!userId) {
      return c.json({ error: "user_id is required" }, 400);
    }

    const db = await getDb();
    const notifications = await db
      .selectFrom("ambient_notifications")
      .select(["id", "message", "priority", "status", "created_at", "delivered_at"])
      .where("user_id", "=", userId)
      .orderBy("created_at", "desc")
      .limit(Number.isFinite(limit) ? limit : 5)
      .execute();

    return c.json({
      notifications: notifications.map((n) => ({
        id: n.id,
        message: n.message,
        priority: n.priority,
        status: n.status,
        created_at: n.created_at ? n.created_at.toISOString() : null,
        delivered_at: n.delivered_at ? n.delivered_at.toISOString() : null,
      })),
    });
  });

  app.get("/notifications/pending", async (c) => {
    const medium = c.req.query("medium");
    if (!medium) {
      return c.json({ error: "medium is required" }, 400);
    }

    const db = await getDb();
    const notifications = await db
      .selectFrom("ambient_notifications")
      .select([
        "id",
        "user_id",
        "target_location",
        "message",
        "priority",
        "routing_reasoning",
        "created_at",
      ])
      .where("target_medium", "=", medium)
      .where("status", "=", "pending")
      .orderBy("priority", "desc")
      .orderBy("created_at", "asc")
      .execute();

    return c.json({
      notifications: notifications.map((n) => ({
        id: n.id,
        user_id: n.user_id,
        target_location: n.target_location,
        message: n.message,
        priority: n.priority,
        routing_reasoning: n.routing_reasoning,
        created_at: n.created_at,
      })),
    });
  });

  app.post("/notifications/recent_unacknowledged", async (c) => {
    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    const userId = payload?.user_id;
    const since = payload?.since;
    if (typeof userId !== "string" || typeof since !== "string") {
      return c.json({ error: "user_id and since are required" }, 400);
    }

    const sinceTime = new Date(since);
    if (Number.isNaN(sinceTime.getTime())) {
      return c.json({ error: "since must be ISO timestamp" }, 400);
    }

    const db = await getDb();
    const notifications = await db
      .selectFrom("ambient_notifications")
      .select([
        "id",
        "message",
        "priority",
        "created_at",
        "delivered_at",
        "status",
        "acknowledged",
        "parent_notification_id",
      ])
      .where("user_id", "=", userId)
      .where("created_at", ">=", sinceTime)
      .where("status", "=", "delivered")
      .where("acknowledged", "=", false)
      .orderBy("created_at", "desc")
      .limit(10)
      .execute();

    return c.json({
      notifications: notifications.map((n) => ({
        id: n.id,
        message: n.message,
        priority: n.priority,
        created_at: n.created_at ? n.created_at.toISOString() : null,
        delivered_at: n.delivered_at ? n.delivered_at.toISOString() : null,
        status: n.status,
        acknowledged: n.acknowledged,
        parent_notification_id: n.parent_notification_id,
      })),
    });
  });

  app.post("/notifications/:notification_id/delivered", async (c) => {
    const notificationId = Number(c.req.param("notification_id"));
    if (!Number.isFinite(notificationId)) {
      return c.json({ error: "Invalid notification_id" }, 400);
    }

    const db = await getDb();
    await db
      .updateTable("ambient_notifications")
      .set({ status: "delivered", delivered_at: nowDate() })
      .where("id", "=", notificationId)
      .execute();

    return c.json({ status: "delivered" });
  });

  app.post("/notifications/:notification_id/acknowledge", async (c) => {
    const notificationId = Number(c.req.param("notification_id"));
    if (!Number.isFinite(notificationId)) {
      return c.json({ error: "Invalid notification_id" }, 400);
    }

    const db = await getDb();
    await db
      .updateTable("ambient_notifications")
      .set({ acknowledged: true, acknowledged_at: nowDate() })
      .where("id", "=", notificationId)
      .execute();

    return c.json({ status: "acknowledged" });
  });

  app.post("/notifications/:notification_id/failed", async (c) => {
    const notificationId = Number(c.req.param("notification_id"));
    if (!Number.isFinite(notificationId)) {
      return c.json({ error: "Invalid notification_id" }, 400);
    }

    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    const errorMessage = typeof payload?.error_message === "string" ? payload.error_message : null;
    if (!errorMessage) {
      return c.json({ error: "error_message is required" }, 400);
    }

    const db = await getDb();
    await db
      .updateTable("ambient_notifications")
      .set({ status: "failed", error_message: errorMessage, delivered_at: nowDate() })
      .where("id", "=", notificationId)
      .execute();

    return c.json({ status: "failed" });
  });
}
